_posts/2024-09-12-v5-release.md

```diff
@@ -0,0 +1,139 @@
+
+#### 3. Name everything. Ordered numerical parameters are not supported.
+
+In Express v4, using regex capture groups you could get numerical parameters (ex. `/user(s?)` => `req.params[0] === 's'`). Now all params must be named. Along with requiring a
+name, we now support all valid JavaScript identifiers or quoted (ex. `/:"this"`).
+
+### Promise Support
+
+This one may be a bit contentious, but I `Promise` we are moving in the right direction. We have added support for returned *rejected* promises from middleware raising errors. This
+*does not include* calling `next` from returned *resolved* promises. There are a lot of edge cases in old express apps around expectations of `Promise` behavior, and before we can
+run we need to walk. For most folks, this means you can now write middleware like the following:
+
+```javascript
+app.use(async (req, res, next) => {
+  req.locals.user = await getUser(req);
+  next();
+});
+```
+
+Notice that we use `async/await` and the `getUser` call may throw (user does not exist, db is down, etc), but we still call `next` if it is successful. We dont need to catch the
+error in line anymore if we want to rely on error handling middleware instead because the router will now catch the rejected promise and treat that as calling `next(err)`.
+
+*editorial note:* Error handling is a huge topic, but one hill I will die on is that errors should be handled as close to the error site as possible. So while this is now handled
+in the router, I would strongly urge you to catch this kind of error within the middleware and handle it without relying on separate error handling middleware.
+
+### Body Parser
+
+We made an assortment of `body-parser` changes:
+
+- Add option to customize the urlencoded body depth with a default value of 32 (@TODO see CVE)
+- Remove deprecated `bodyParser()` combination middleware
+- `req.body` is no longer always initialized to `{}`
+- `urlencoded` parser now defaults `extended` to false
+- Added brotli support
+
+### Removed APIs
+
+We have removed a bunch of apis which were primarily from v3.
+
+- `res.redirect('back')` and `res.location('back')` are no longer a supported magic string, explicitly use `req.get('Referrer') || '/'`
+- `res.send(status, body)` - use `res.status(status).send(body)`
+- `res.redirect(url, status)` - use `res.redirect(status, url)`
+- `res.jsonp(status, obj)` - use `res.status(status).jsonp(obj)`
+- `res.json(status, obj)` - use `res.status(status).json(obj)`
+- `app.param(fn)` - use `req.params`, `req.body` or `req.query` instead
+- `app.del` - use `app.delete`
+- `req.acceptsCharset` - use `req.acceptsCharsets`
+- `req.acceptsEncoding` - use `req.acceptsEncodings`
+- `req.acceptsLanguage` - use `req.acceptsLanguages`
+- `res.json(obj, status)` signature - use `res.json(status, obj)`
+- `res.jsonp(obj, status)` signature - use `res.jsonp(status, obj)`
+- `res.send(body, status)` signature - use `res.send(status, body)`
```

Comment:
I think I have a understanding gap here, please help me to understand. Line 129 says,

```md
`res.send(body, status)` signature - use `res.send(status, body)`
```

But line 119 says,
```md
`res.send(status, body)` - use `res.status(status).send(body)`
```

And since `res.send(status, body)` no longer works, it might be confusing to suggest that? Sames goes for other few methods as I see.

Reply:
This is a good callout. It was pulled from the changelog which evolved as the betas did. These are deprecated signatures. `res.send()` no longer accepts a status input.